src/vsr/replica.zig

```diff
@@ -2211,13 +2228,34 @@ pub fn ReplicaType(
         // is the op-number. Then they execute all operations known to be committed that they
         // haven’t executed previously, advance their commit number, and update the information in
         // their client table.
-        fn on_start_view(self: *Replica, message: *const Message.StartView) void {
+        fn on_start_view(self: *Replica, message: *Message.StartView) void {
             assert(message.header.command == .start_view);
             if (self.ignore_view_change_message(message.base_const())) return;
 
+            if (self.status == .recovering_head) {
+                if (message.header.view > self.view or
+                    message.header.op >= self.op_prepare_max() or
+                    message.header.nonce == self.nonce)
+                {
+                    // This SV is guaranteed to have originated after the replica crash,
+                    // it is safe to use to determine the head op.
+                } else {
+                    log.mark.debug(
+                        "{}: on_start_view: ignoring (recovering_head, nonce mismatch)",
+                        .{self.replica},
+                    );
+                    return;
+                }
+            }
```

Comment:
I pulled the `nonce mismatch` check up out of `ignore_view_change_message()` because otherwise:

1. While in `status=recovering_head`, receive a SV message from a newer view (with nonce=0).
2. Since `message.header.view > self.view` we accept it and set `self.view = message.header.view`.
3. Then we call `on_start_view_set_journal(message)` (still with that same SV).
4. Only now the views match, so we require that the nonces match as well -- we panic on `assert(!self.ignore_view_change_message(message.base_const()))`.